### DISCUSSION TITLE
Request an example of the datamodel-path-selector option in the retrieve command

### DISCUSSION THREAD

**Discussion by Jennyism** [filecoin-project/boost#1951] [Client]
I'm exploring using boost for retrieve but I'm having some issues. I read `https://ipld.io/docs/data-model/` documentation to try to understand the data-model-path-selector flag, but it doesn't feel like it helps.
I tried using the `/ls` expression (boost retrieve --datamodel-path-selector /ls -p <provider> <data cid>), and although the program threw panic, I could tell from the logs that it received 13KB of data. This is consistent with the size of the data I retrieved using Lassie. 

My need is to get a more standard example of the `datamodel-path-selector`, and I want to confirm if the failure of my retrieval method using \ls is a code issue and not a problem with my usage. 

Below is the error report and the associated source code: 

> RETRIEVAL STATS (FIL)
> -----
>
> Size:          13451 (13 KiB)
> Duration:      289.294336ms
> Average Speed: 46495 (45 KiB/s)
> Ask Price:     0 (0 FIL)
> Total Payment: 0 (0 FIL)
> Num Payments:  0
> Peer:          *****
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x2de0146]
>
> goroutine 1 [running]:
> main.init.func13.2({0xc0008a1180, 0x0, {{0x0, 0x0, 0x0}}, {{{0x0, 0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
>         /root/lotus_compiler/boost/cmd/boost/retrieve_cmd.go:250 +0x106
> github.com/ipld/go-ipld-prime/traversal.Progress.visit({0xc0008a1180, 0x0, {{0x0, 0x0, 0x0}}, {{{0x0, 0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
>         /root/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.21.0/traversal/walk.go:347 +0xc3
> github.com/ipld/go-ipld-prime/traversal.Progress.walkAdv({0xc0008a1180, 0x0, {{0x0, 0x0, 0x0}}, {{{0x0, 0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
>         /root/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.21.0/traversal/walk.go:210 +0x1f7
> github.com/ipld/go-ipld-prime/traversal.Progress.walkBlock({0xc0008a1180, 0x0, {{0x0, 0x0, 0x0}}, {{{0x0, 0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
>         /root/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.21.0/traversal/walk.go:180 +0x10f
> github.com/ipld/go-ipld-prime/traversal.Progress.WalkAdv({0xc0008a1180, 0x0, {{0x0, 0x0, 0x0}}, {{{0x0, 0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
>         /root/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.21.0/traversal/walk.go:160 +0x15e
> github.com/filecoin-project/boost/markets/utils.TraverseDag({0x7f1acc8, 0xc000343bd0}, {0x7f24c08, 0xc000711698}, {{0xc00017a120?, 0x0?}}, {0x7f3d9c0?, 0xc00033d7c0?}, 0xc00092d8d8)
>         /root/lotus_compiler/boost/markets/utils/selectors.go:93 +0x445
> main.init.func13(0xc000128c80)
>         /root/lotus_compiler/boost/cmd/boost/retrieve_cmd.go:236 +0x1134
> github.com/urfave/cli/v2.(*Command).Run(0xa744ae0, 0xc000128c80, {0xc00027eba0, 0x6, 0x6})
>         /root/go/pkg/mod/github.com/urfave/cli/v2@v2.27.2/command.go:276 +0x97d
> github.com/urfave/cli/v2.(*Command).Run(0xc000171760, 0xc000128a00, {0xc000052230, 0x7, 0x7})
>         /root/go/pkg/mod/github.com/urfave/cli/v2@v2.27.2/command.go:269 +0xbb7
> github.com/urfave/cli/v2.(*App).RunContext(0xc000203400, {0x7f1ac58, 0xa9001e0}, {0xc000052230, 0x7, 0x7})
>         /root/go/pkg/mod/github.com/urfave/cli/v2@v2.27.2/app.go:333 +0x58b
> github.com/urfave/cli/v2.(*App).Run(...)
>         /root/go/pkg/mod/github.com/urfave/cli/v2@v2.27.2/app.go:307
> main.main()
>         /root/lotus_compiler/boost/cmd/boost/main.go:51 +0x3d5

```
func(p traversal.Progress, n ipld.Node, r traversal.VisitReason) error {
	if r == traversal.VisitReason_SelectionMatch {

		if p.LastBlock.Path.String() != p.Path.String() {
			return xerrors.Errorf("unsupported selection path '%s' does not correspond to a node boundary (a.k.a. CID link)", p.Path.String())
		}

		cidLnk, castOK := p.LastBlock.Link.(cidlink.Link)
		if !castOK {
			return xerrors.Errorf("cidlink cast unexpectedly failed on '%s'", p.LastBlock.Link.String())
		}

		c = cidLnk.Cid
		subRootFound = true
	}
	return nil
},
```

**Comment by LexLuthr** [comment]
You are definitely using wrong selector there. IPLD selector are not like you commands. Not sure what gave you that impression.
https://github.com/filecoin-project/lotus/commit/469960ce0ef1f26b10ec4569e7c8b5832291de32#diff-193d7ce48ff1005390415baf8032ce14bf59a5cd1a85ea65a2adc078e9997931L408 is an old example of using selectors to list objects. Can you tell me why you want to use selectors?

**Reply by Jennyism** [reply]
I was exploring a way for retrieve like the "lotus client ls" command, and when I saw that it used a JSON selector in the source code(`{"R":{"l":{"depth":%d},":>":{"a":{">":{"|":[{"@":{}},{".":{}}]}}}}}`), I guessed that it could also be implemented in boost, but I only got a panic. So I started exploring the selector, but I couldn't find a suitable example. 

Rather, I was looking for a way to quickly verify that the SP can be retrieved. I think it would be more convenient and effective to be able to use `ls` for verification than to use `retrieve`. I was able to do a random spot check based on the node information returned by `ls`, while `retrieve` can only download all the retrieved data.
